rpc/export.go

```diff
@@ -154,26 +155,71 @@ func (c *lockedConn) sendCap(d rpccp.CapDescriptor, client capnp.Client) (_ expo
 	state.Metadata.Lock()
 	defer state.Metadata.Unlock()
 	id, ok := c.findExportID(state.Metadata)
+	var ee *expent
 	if ok {
-		ent := c.lk.exports[id]
-		ent.wireRefs++
-		d.SetSenderHosted(uint32(id))
-		return id, true, nil
+		ee = c.lk.exports[id]
+		ee.wireRefs++
+	} else {
+		// Not already present; allocate an export id for it:
+		ee = &expent{
+			client:    client.AddRef(),
+			wireRefs:  1,
+			isPromise: state.IsPromise,
+		}
+		id = exportID(c.lk.exportID.next())
+		if int64(id) == int64(len(c.lk.exports)) {
+			c.lk.exports = append(c.lk.exports, ee)
+		} else {
+			c.lk.exports[id] = ee
+		}
+		c.setExportID(state.Metadata, id)
 	}
+	if ee.isPromise {
+		// Send a promise, wait for the resolution asynchronously, then send
```

Comment:
Could we abstract this block in a separate function?  `sendCap` is already quite long, and I think it would greatly improve readability.  I'd also be fine with just abstracting the anonymous goroutine into a separate (named) function.

Reply:
I agree in principle, though it's not obvious to me how to split this up in a way that's conceptually coherent; I'll have to think on it.

Reply:
Ok, split most of this out into a `sendSenderPromise` method.